Changes in module 'nf-core/trimgalore'
--- modules/nf-core/trimgalore/main.nf
+++ modules/nf-core/trimgalore/main.nf
@@ -72,4 +72,29 @@
         END_VERSIONS
         """
     }
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    if (meta.single_end) {
+        """
+        touch ${prefix}.50bp_5prime.fq.gz
+
+        cat <<-END_VERSIONS > versions.yml
+        "${task.process}":
+            trimgalore: \$(echo \$(trim_galore --version 2>&1) | sed 's/^.*version //; s/Last.*\$//')
+            cutadapt: \$(cutadapt --version)
+        END_VERSIONS
+        """
+    } else {
+        """
+        touch ${prefix}_1.50bp_5prime.fq.gz
+        touch ${prefix}_2.50bp_5prime.fq.gz
+
+        cat <<-END_VERSIONS > versions.yml
+        "${task.process}":
+            trimgalore: \$(echo \$(trim_galore --version 2>&1) | sed 's/^.*version //; s/Last.*\$//')
+            cutadapt: \$(cutadapt --version)
+        END_VERSIONS
+        """
+    }
 }

************************************************************
