Changes in module 'nf-core/samtools/index'
--- modules/nf-core/samtools/index/main.nf
+++ modules/nf-core/samtools/index/main.nf
@@ -11,10 +11,8 @@
     tuple val(meta), path(input)
 
     output:
-    tuple val(meta), path("*.bai") , optional:true, emit: bai
-    tuple val(meta), path("*.csi") , optional:true, emit: csi
-    tuple val(meta), path("*.crai"), optional:true, emit: crai
-    path  "versions.yml"           , emit: versions
+    tuple val(meta), path("*.{bai,crai}") , emit: index
+    path  "versions.yml"                  , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -37,7 +35,6 @@
     stub:
     """
     touch ${input}.bai
-    touch ${input}.crai
     touch ${input}.csi
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
